classes/author/Repository.inc.php

```diff
@@ -0,0 +1,228 @@
+            $errors = $schemaService->formatValidationErrors($validator->errors(), $schemaService->get(PKPSchemaService::SCHEMA_AUTHOR), $allowedLocales);
+        }
+
+        HookRegistry::call('Author::validate', [&$errors, $author, $props, $allowedLocales, $primaryLocale]);
+
+        return $errors;
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::add()
+     */
+    public function add(Author $author): int
+    {
+        $authorId = $this->dao->insert($author);
+        $author = Repo::author()->get($authorId);
+
+        HookRegistry::call('Author::add', [&$author]);
+
+        return $author->getId();
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::edit()
+     */
+    public function edit(Author $author, array $params): Author
+    {
+        $newAuthor = Repo::author()->newDataObject(array_merge($author->_data, $params));
+
+        HookRegistry::call('Author::edit', [&$newAuthor, $author, $params, $this->request]);
+
+        $this->dao->update($newAuthor);
+
+        $newAuthor = Repo::author()->get($newAuthor->getId());
+
+        return $newAuthor;
+    }
+
+    /**
+     * @copydoc \PKP\services\entityProperties\EntityWriteInterface::delete()
+     */
+    public function delete(Author $author)
+    {
+        HookRegistry::call('Author::delete::before', [&$author]);
+        $this->dao->delete($author);
+        HookRegistry::call('Author::delete', [&$author]);
+    }
+
+    /**
+     * Update author names when publication locale changes.
+     *
+     * @param $publicationId int
+     * @param $oldLocale string
+     * @param $newLocale string
+     */
+    public function changePublicationLocale($publicationId, $oldLocale, $newLocale)
+    {
+        $authors = $this->getMany(
+            $this
+                ->getCollector()
+                ->filterByPublicationIds([$publicationId])
+        );
+
+        foreach ($authors as $author) {
+            if (empty($author->getGivenName($newLocale))) {
+                if (empty($author->getFamilyName($newLocale)) && empty($author->getPreferredPublicName($newLocale))) {
+                    // if no name exists for the new locale
+                    // copy all names with the old locale to the new locale
+                    $author->setGivenName($author->getGivenName($oldLocale), $newLocale);
+                    $author->setFamilyName($author->getFamilyName($oldLocale), $newLocale);
+                    $author->setPreferredPublicName($author->getPreferredPublicName($oldLocale), $newLocale);
+                } else {
+                    // if the given name does not exist, but one of the other names do exist
+                    // copy only the given name with the old locale to the new locale, because the given name is required
+                    $author->setGivenName($author->getGivenName($oldLocale), $newLocale);
+                }
+
+                $this->edit($author, []);
```

Comment:
In this case, you can use the dao directly. That's because the edit is an automated change based on another action in the system, and you already have the author object modified:

```
$this->dao->update($author);
```

Reply:
I did that so that the hooks could be called. 

Changed.